src/main/java/com/exasol/bucketfs/uploadnecassity/ChecksumUploadNecessityCheckStrategy.java

```diff
@@ -0,0 +1,144 @@
+            final String fileName = parts[parts.length - 1];
+            final List<String> filesInBucketDirectory = bucket.listContents(getDirectory(parts));
+            if (Files.size(file) > ONE_MEGABYTE && filesInBucketDirectory.contains(fileName)) {
+                return !localMd5Checksum(file).equals(getMd5Checksum(fullFileNameInBucketFs, bucket));
+            } else {
+                return true;
+            }
+        } catch (final BucketAccessException | NoSuchAlgorithmException | IOException exception) {
+            throw new BucketAccessException(ExaError.messageBuilder("E-BFSJ-17")
+                    .message("Failed to check if we need to upload {{file}}.", fullFileNameInBucketFs).toString(),
+                    exception);
+        }
+    }
+
+    private String getDirectory(final String[] parts) {
+        final String directory = Arrays.stream(parts).limit(parts.length - 1)
+                .collect(Collectors.joining(BucketConstants.PATH_SEPARATOR));
+        if (directory.startsWith(BucketConstants.PATH_SEPARATOR)) {
+            return directory.substring(1);
+        } else {
+            return directory;
+        }
+    }
+
+    private String localMd5Checksum(final Path localPath) throws NoSuchAlgorithmException, IOException {
+        final MessageDigest checksumBuilder = MessageDigest.getInstance("MD5");
+        try (final InputStream inputStream = Files.newInputStream(localPath);
+                final DigestInputStream checksumBuildingStream = new DigestInputStream(inputStream, checksumBuilder)) {
+            final byte[] buffer = new byte[1000];
+            while (checksumBuildingStream.read(buffer) != -1) {
+                // nothing to do. Just read to run through the stream.
+            }
+        }
+        return toHex(checksumBuilder.digest());
+    }
+
+    /**
+     * Get the MD5-checksum of a file in BucketFS.
+     *
+     * @param fileInBucketFs path to a file in bucketFs
+     * @param bucket         bucket the file is stored in
+     * @return md5 checksum
+     * @throws BucketAccessException if checksum calculation failed
+     */
+    public String getMd5Checksum(final String fileInBucketFs, final ReadOnlyBucket bucket)
+            throws BucketAccessException {
+        installChecksumUdf();
+        try (final PreparedStatement statement = this.sqlConnection
+                .prepareStatement("SELECT " + UDF_FULL_NAME + "(?)")) {
+            statement.setString(1,
+                    "/buckets/" + bucket.getBucketFsName() + "/" + bucket.getBucketName() + "/" + fileInBucketFs);
+            try (final ResultSet result = statement.executeQuery()) {
+                result.next();
+                return result.getString(1);
+            }
+        } catch (final SQLException exception) {
+            throw new BucketAccessException(ExaError.messageBuilder("F-BFSJ-15")
+                    .message("Failed to determine checksum of file {{file}} in BucketFs using a python UDF.",
```

Comment:
```suggestion
                    .message("Failed to determine checksum of file {{file}} in BucketFS using a UDF {{udf}}.",
```
Provide database object ID.